src/Libraries/Microsoft.Extensions.AI/ChatCompletion/AnonymousChatClient.cs

```diff
@@ -0,0 +1,242 @@
+
+    /// <summary>
+    /// Initializes a new instance of the <see cref="AnonymousChatClient"/> class.
+    /// </summary>
+    /// <param name="innerClient">The inner generator.</param>
+    /// <param name="services">Services associated with the chat client.</param>
+    /// <param name="sharedFunc">
+    /// A delegate that provides the implementation for both <see cref="CompleteAsync"/> and <see cref="CompleteStreamingAsync"/>.
+    /// In addition to the arguments for the operation, it's provided with a delegate to the inner client that should be
+    /// used to perform the operation on the inner client. It will handle both the non-streaming and streaming cases.
+    /// </param>
+    /// <remarks>
+    /// This overload may be used when the anonymous implementation needs to provide pre- and/or post-processing, but doesn't
+    /// need to interact with the results of the operation, which will come from the inner client.
+    /// </remarks>
+    /// <exception cref="ArgumentNullException"><paramref name="innerClient"/> is <see langword="null"/>.</exception>
+    /// <exception cref="ArgumentNullException"><paramref name="sharedFunc"/> is <see langword="null"/>.</exception>
+    public AnonymousChatClient(
+        IChatClient innerClient,
+        IServiceProvider? services,
+        Func<Arguments, Func<IList<ChatMessage>, ChatOptions?, CancellationToken, Task>, Task> sharedFunc)
+        : base(innerClient)
+    {
+        _ = Throw.IfNull(innerClient);
+        _ = Throw.IfNull(sharedFunc);
+
+        _services = services;
+        _sharedFunc = sharedFunc;
+    }
+
+    /// <summary>
+    /// Initializes a new instance of the <see cref="AnonymousChatClient"/> class.
+    /// </summary>
+    /// <param name="innerClient">The inner generator.</param>
+    /// <param name="services">Services associated with the chat client.</param>
+    /// <param name="completeFunc">
+    /// A delegate that provides the implementation for <see cref="CompleteAsync"/>. When <see langword="null"/>,
+    /// <paramref name="completeStreamingFunc"/> must be non-null, and the implementation of <see cref="CompleteAsync"/>
+    /// will use <paramref name="completeStreamingFunc"/> for the implementation.
+    /// </param>
+    /// <param name="completeStreamingFunc">
+    /// A delegate that provides the implementation for <see cref="CompleteStreamingAsync"/>. When <see langword="null"/>,
+    /// <paramref name="completeFunc"/> must be non-null, and the implementation of <see cref="CompleteStreamingAsync"/>
+    /// will use <paramref name="completeFunc"/> for the implementation.
+    /// </param>
+    /// <exception cref="ArgumentNullException"><paramref name="innerClient"/> is <see langword="null"/>.</exception>
+    /// <exception cref="ArgumentNullException">Both <paramref name="completeFunc"/> and <paramref name="completeStreamingFunc"/> are <see langword="null"/>.</exception>
+    public AnonymousChatClient(
+        IChatClient innerClient,
```

Comment:
Unless I'm misunderstanding something, the only reason we require an inner client is to drive the shared func, which isn't a mode being exercised at all by the particular constructor.

In general, I think requiring a inner client negates the purpose of exposing an anonymous interface implementation.

Reply:
Thanks, your example makes it clearer to me what this type represents. Effectively it's modelling an aspnetcore-style middleware? I interpreted the term "anonymous client" as a helper that just that creates thin anonymous objects using delegates.

> No, it's necessary to be passed to the base because it's used by the other delegating members, i.e.. Metadata, GetService, and Dispose.

I can see how it accelerates the definition of those ancillary methods, but not needing to specify an underlying client feels like goodness too. Some of these ancillary members could be specified using additional parameters.

My overall feeling is that the two constructors are different enough that they should be factored into separate types with different names.